site/content/posts/2023-08-09-v0.25.0-impersonation-proxy-with-external-certs.md

```diff
@@ -0,0 +1,337 @@
+
+```shell
+$ cat << EOF > contour-ingress-impersonation-proxy.yaml
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: impersonation-proxy
+  namespace: pinniped-concierge
+spec:
+  virtualhost:
+    fqdn: impersonation-proxy-mtls.local
+    tls:
+      passthrough: true
+  tcpproxy:
+    services:
+    - name: pinniped-concierge-impersonation-proxy-cluster-ip
+      port: 443
+EOF
+
+$ kubectl apply \
+    --filename contour-ingress-impersonation-proxy.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Now generate the Pinniped kubeconfig so that you can perform mTLS with the impersonation proxy.
+
+Note that using a static-token does embed those credentials into your kubeconfig.
+Never use `local-user-authenticator` in production.
+
+```shell
+# add 127.0.0.1 impersonation-proxy-mtls.local to your /etc/hosts!
+$ pinniped get kubeconfig \
+    --static-token "pinny:password123" \
+    --concierge-authenticator-type webhook \
+    --concierge-authenticator-name local-user-authenticator \
+    --concierge-mode ImpersonationProxy \
+    --kubeconfig kind-with-contour.kubeconfig.yaml \
+    > pinniped-kubeconfig.yaml
+```
+
+Now perform an action as user pinny!
+
+```shell
+$ kubectl get pods -A \
+    --kubeconfig pinniped-kubeconfig.yaml
+Error from server (Forbidden): pods is forbidden: User "pinny" cannot list resource "pods" in API group "" at the cluster scope: decision made by impersonation-proxy.concierge.pinniped.dev
+```
+
+This does result in an error because the cluster does not have any `RoleBindings` or `ClusterRoleBindings` that allow your user pinny or the group `group-for-mtls` to perform any actions on the cluster.
+Let’s make a `ClusterRoleBinding` that grants this group cluster admin privileges.
+
+```shell
+# Perform this as the cluster admin using the kind kubeconfig
+$ kubectl create clusterrolebinding mtls-admins \
+    --clusterrole=cluster-admin \
+    --group=group-for-mtls \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+# Now try again with the Pinniped kubeconfig
+$ kubectl get pods -A \
+    --kubeconfig pinniped-kubeconfig.yaml
+NAMESPACE                  NAME                                                      READY   STATUS      RESTARTS       AGE
+pinniped-concierge         pinniped-concierge-f4c78b674-bt6zl                        1/1     Running     0              3h36m
+```
+
+Congratulations, you have successfully performed mTLS authentication between your local client (kubectl, using the pinniped CLI) and the impersonation proxy inside the cluster.
+
+To verify that your username and groups are visible to Kubernetes, run the `pinniped whoami` command.
+
+```shell
+pinniped whoami \
+    --kubeconfig pinniped-kubeconfig.yaml
+```
+
+Now, to verify that the generated kubeconfig `pinniped-kubeconfig.yaml` has the contents of `self-signed-ca-for-kind-testing.pem.b64` as the contained CA bundle for the cluster,
```

Comment:
Finally, lets review service endpoints and expected certificate values.

The endpoints are as follows:
```
# your pinniped-kubeconfig fill contain the impersonation proxy endpoint
# indicating that all of your kubectl requests will flow through the impersonation proxy
cat pinniped-kubeconfig.yaml | yq ".clusters[0].cluster.server"
# https://impersonation-proxy-mtls.local/

# verify the impersonation proxy endpoint in the HTTPProxy resource
kubectl get HTTPProxy impersonation-proxy -n pinniped-concierge -o jsonpath="{.spec.virtualhost.fqdn}"
# impersonation-proxy-mtls.local

# the CredentialIssuer configures the impersonationProxy to respond to this endpoint
kubectl get CredentialIssuer pinniped-concierge-config -o jsonpath="{.spec.impersonationProxy.externalEndpoint}"
# impersonation-proxy-mtls.local

# the webhook authenticator supports responds at this endpoint
kubectl get WebhookAuthenticator local-user-authenticator -n local-user-authenticator -o jsonpath="{.spec.endpoint}"
#  https://local-user-authenticator.local-user-authenticator.svc/authenticate
```
And the certificates:

```bash
# we provided this CA to the impersonation proxy
kubectl get CredentialIssuer pinniped-concierge-config -o jsonpath="{.spec.impersonationProxy.tls.certificateAuthorityData}"

# the impersonation proxy should advertise this CA certificate in its status
kubectl get CredentialIssuer pinniped-concierge-config -o jsonpath="{.status.strategies[1].frontend.impersonationProxyInfo.certificateAuthorityData}"

# and this CA certificate should appear in your pinniped generated kubeconfig
cat pinniped-kubeconfig.yaml | yq ".clusters[0].cluster.certificate-authority-data"
```

Reply:
So then, the `pinniped-kubeconfig.yaml` file actually matches the `CredentialIssuer.impersonationProxy`'s CA:


```bash
cat pinniped-kubeconfig.yaml | yq ".clusters[0].cluster.certificate-authority-data"

kubectl get CredentialIssuer pinniped-concierge-config -o jsonpath="{.status.strategies[1].frontend.impersonationProxyInfo.certificateAuthorityData}"
```
but do not match the CA from the WebHookAuthenticator

```bash
kubectl get WebhookAuthenticator local-user-authenticator -n local-user-authenticator -o jsonpath="{.spec.certificateAuthorityData}"
```

Reply:
Since `self-signed-ca-for-kind-testing.pem.b64` is applied to `credentialissuer.spec.impersonationProxy.tls.certificateAuthorityData` but is originally from the `Secret` called `self-signed-ca-for-kind-testing`  which represents the `cert-manager` `Certificate` called `my-selfsigned-ca`, 

So working backwards:
- `self-signed-ca-for-kind-testing.pem.b64` is applied to `credentialissuer.spec.impersonationProxy.tls.certificateAuthorityData`
- it is originally from the `Secret` called `self-signed-ca-for-kind-testing`
- This is from the `ClusterIssuer` called `my-ca-issuer` which has a `spec.ca.secretName` where it references a CA certificate
- which is the `Certificate` that is called `my-selfsigned-ca` that has `.spec.secretName: self-signed-ca-for-kind-testing` as well.

Reply:
So `cat self-signed-ca-for-kind-testing.pem.b64` == `cat pinniped-kubeconfig.yaml | yq ".clusters[0].cluster.certificate-authority-data"`

Reply:
Resolving locally.